src/query/automaton_weight.rs

```diff
@@ -1,80 +1,78 @@
 use crate::core::SegmentReader;
-use crate::query::ConstScorer;
-use crate::query::{BitSetDocSet, Explanation};
+use crate::query::fuzzy_query::DfaWrapper;
+use crate::query::score_combiner::SumCombiner;
+use crate::query::Explanation;
+use crate::query::{ConstScorer, Union};
 use crate::query::{Scorer, Weight};
 use crate::schema::{Field, IndexRecordOption};
-use crate::termdict::{TermDictionary, TermStreamer};
+use crate::termdict::{TermDictionary, TermWithStateStreamer};
 use crate::TantivyError;
 use crate::{DocId, Score};
-use common::BitSet;
+use std::any::{Any, TypeId};
 use std::io;
 use std::sync::Arc;
 use tantivy_fst::Automaton;
 
 /// A weight struct for Fuzzy Term and Regex Queries
-pub struct AutomatonWeight<A> {
+pub struct AutomatonWeight<A, F> {
```

Comment:
The automaton weight is for regex and fuzzy queries, as mentionned above.

F has no meaning for regex queries.

Reply:
>  F has no meaning for regex queries.

In this case, it can default to some no-op function like `|_| 1.0`

>  We can have F match F: Fn(A::State) -> Score .

I don't really understand this because there would still be an F for a regex query then, wouldn't there?

> For better usability, we can also offer a default to the Const function that returns one.

If `F` is a `Fn` then we cannot provide `struct AutomatonWeight<A, F = <some default>>` because we can't name `|_| 1.0`, nor implement it for a `struct DefaultScorer;` because that's unstable

Reply:
I also don't quite understand what you mean by "we want the user to pass an array over the distance, not a function of the state", if you are also suggesting `Fn(A::State) -> Score`. Apologies if I am misinterpreting this

Reply:
```
I don't really understand this because there would still be an F for a regex query then, wouldn't there?
```
Ah yes... :)
But there is one big notable difference.

The role of F can then be explained without necessary talking about the fuzzy term query.

```
 because we can't name |_| 1.0
```

> nor implement it for a struct DefaultScorer; because that's unstable

Which feature is unstable exactly?

> I also don't quite understand what you mean by "we want the user to pass an array over the distance, not a function of the state", if you are also suggesting Fn(A::State) -> Score. Apologies if I am misinterpreting this

We can keep the function.
The user would give the FuzzyTermQuery an array over the distance rather than a function. It's tiny anyway.

user gives `[f32; max_distance + 1]`
`FuzzyTermQuery` takes that an transforms into a `Fn(u32) -> f32`,
The implementation take a state, uses the automaton to get a distance, and do a lookup in the user array.

Reply:
> Which feature is unstable exactly?}

Implementing a Fn trait is unstable: https://github.com/rust-lang/rust/issues/29625. This can be mitigated by making `trait AutomatonScorer { fn score(A::State) -> f32  }` or similar.

>  The user would give the FuzzyTermQuery an array over the distance rather than a function. It's tiny anyway.

So, in `FuzzyTermQuery` the user provides a function, and then at the AutomatonWeight boundary it is converted to a suitable `F`?